src/main/java/de/tum/in/www1/artemis/domain/ShortAnswerQuestion.java

```diff
@@ -122,44 +159,316 @@ public void setCorrectMappings(List<ShortAnswerMapping> shortAnswerMappings) {
-    }
+        // check if at least one correct mapping exists
+        return getCorrectMappings() != null && !getCorrectMappings().isEmpty();
 
-    @Override
-    public String toString() {
-        return "ShortAnswerQuestion{" +
-            "id=" + getId() +
-            "}";
+        // TODO (?): Add checks for "is solvable" and "no misleading correct mapping" --> look at the implementation in the client
     }
 
+     /**
+      * Get all solution items that are mapped to the given spot
+      *
+      * @param spot the spot  we want to find the correct solutions for
+      * @return all solutions that are defined as correct for this spot
+      */
+     public Set<ShortAnswerSolution> getCorrectSolutionForSpot(ShortAnswerSpot spot) {
+         Set<ShortAnswerSolution> result = new HashSet<>();
+         for (ShortAnswerMapping mapping : correctMappings) {
+             if (mapping.getSpot().equals(spot)) {
+                 result.add(mapping.getSolution());
+             }
+         }
+         return result;
+     }
+
+     /**
+      * Get solution by ID
+      *
+      * @param solutionId the ID of the solution, which should be found
+      * @return the sollution with the given ID, or null if the solution is not contained in this question
+      */
+     public ShortAnswerSolution findSolutionById (Long solutionId) {
+
+         if (solutionId != null) {
+             // iterate through all dragItems of this quiz
+             for (ShortAnswerSolution solution : solutions) {
+                 // return dragItem if the IDs are equal
+                 if (solution.getId().equals(solutionId)) {
+                     return solution;
+                 }
+             }
+         }
+         return null;
+     }
+
+     /**
+      * Get spot by ID
+      *
+      * @param spotId the ID of the spot, which should be found
+      * @return the spot with the given ID, or null if the spot is not contained in this question
+      */
+     public ShortAnswerSpot findSpotById (Long spotId) {
+
+         if (spotId != null) {
+             // iterate through all dropLocations of this quiz
+             for (ShortAnswerSpot spot : spots) {
+                 // return dropLocation if the IDs are equal
+                 if (spot.getId().equals(spotId)) {
+                     return spot;
+                 }
+             }
+         }
+         return null;
+     }
+
+
+     /**
+      * undo all solution- and spot-changes which are not allowed ( adding them)
+      *
+      * @param originalQuestion the original Question-object, which will be compared with this question
+      *
+      */
     @Override
     public void undoUnallowedChanges(Question originalQuestion) {
-        //TODO Francisco implement
+        if (originalQuestion instanceof ShortAnswerQuestion) {
+            ShortAnswerQuestion saOriginalQuestion = (ShortAnswerQuestion) originalQuestion;
+            //undo unallowed solution changes
```

Comment:
Comments aren't needed when code is autoesplicative, try to avoid trivial comments :)

Writing good comments is indeed an hard task, and what "good" means is anyway up to debate. I suggest reading this guide, that personally I find very well-written and opinionated. It's from the author of Redis.

http://antirez.com/news/124

Reply:
cheers man! :)